src/main/java/vg/civcraft/mc/civmodcore/serialization/NBTSerialization.java

```diff
@@ -0,0 +1,160 @@
+                    continue;
+                }
+				System.out.println("NBTSerializable[" + clazz.getName() + "] alias [" + alias + "] registered.");
+                registeredClasses.putIfAbsent(alias, clazz);
+            }
+        }
+    }
+
+    /**
+	 * Unregisters an {@link NBTSerializable} class and all its paths.
+	 *
+	 * @param clazz The NBTSerializable class to unregister.
+	 *
+	 * @throws IllegalArgumentException Throws if the given class to unregister is null.
+	 * */
+    public static <T extends NBTSerializable> void unregisterNBTSerializable(Class<T> clazz) {
+    	if (clazz == null) {
+			throw new IllegalArgumentException("Cannot unregister NBTSerializable: " +
+					"the given class is null.");
+		}
+		System.out.println("NBTSerializable[" + clazz.getName() + "] unregistered.");
+        registeredClasses.values().remove(clazz);
+    }
+
+    /**
+	 * Unregisters all {@link NBTSerializable} classes.
+	 * */
+    public static void clearAllRegistrations() {
+    	System.out.println("All NBTSerialization registered classes have been cleared.");
+    	registeredClasses.clear();
+	}
+
+	/**
+	 * Serializes an {@link NBTSerializable} class into an {@link NBTCompound} for the purpose of being able to
+	 * deserialize it later. For this purpose, {@link NBTSerializable#serialize(NBTCompound) NBTSerializable's
+	 * serialise} method should not be called directly.
+	 *
+	 * @param serializable The {@link NBTSerializable} class to serialize.
+	 * @return Returns an {@link NBTCompound} with the serialised data, or null if the given object was null, or if an
+	 * error occurred during serialization.
+	 * */
+    public static <T extends NBTSerializable> NBTCompound serialize(T serializable) {
+        if (serializable == null) {
+            return null;
+        }
+        NBTCompound nbt = new NBTCompound();
+		try {
+			serializable.serialize(nbt);
+		}
+		catch (Exception exception) {
+			exception.printStackTrace();
+			return null;
+		}
+		nbt.setString(NBT_CLASS_PATH_KEY, serializable.getClass().getName());
+        return nbt;
+    }
+
+    /**
+	 * Deserializes an {@link NBTCompound} into a registered {@link NBTSerializable} class, if it exists.
+	 *
+	 * @param <T> The class type that will be returned. It MUST match the class that was registered exactly.
+	 * @param nbt The NBT data that's used to populate the new instance of the class.
+	 * @return Returns a populated instance of a class, or null. It will return null if 1) the given NBTCompound is null
+	 * or empty, 2) the class identifier cannot be found, 3) the class cannot be matched, or 4) there was an error in
+	 * casting the instance into the given generic type.
+	 * */
+	@SuppressWarnings("unchecked")
+    public static <T extends NBTSerializable> T deserialize(NBTCompound nbt) {
+        if (nbt == null || nbt.isEmpty()) {
+            return null;
+        }
+        String path = nbt.getString(NBT_CLASS_PATH_KEY);
+        nbt.remove(NBT_CLASS_PATH_KEY);
```

Comment:
Neither do I think its a good decision design to have a deserialize method destroy the binary object in the process, nor do I think it should do so without documenting that.

Reply:
The serialize method adds the class path, and the deserialize method removes it. The idea is to separate the class from the classpath process entirely, that the class can and should only know about the data itself.

Reply:
Yes, but the problem there is that a use case might want to deserialize the same binary data multiple times, which wouldn't work with this.

Reply:
That's... a fair point

Reply:
After thinking about it for a while, I've decided to keep the `.remove()`, as I still think that the class being deserialised into should be ignorant of the deserialisation process itself. But you raised a good point that this should not destructive, especially if you wish to deserialise the same data multiple times, and so this is resolved by giving the class a copy of the data, as you don't want deserialisers to potentially corrupt the data for proceeding deserialisations.